### PR TITLE
space - Yieni & Cathy

### PR DESCRIPTION
**Pull request by yieknee**
# Assignment Submission: OO Ride Share

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
:------------- | :-------------
How did getting up to speed on the existing codebase go? If it went well, what worked? If it didn't go well, what will you do differently next time? |  We read through each class to see if we understood what was happening. What didn't work was starting at the abstract class. We also visually mapped out the interactions.
What _inheritance_ relations exist between classes? | Driver, Passenger, and Trip are subclasses of CSVRecord.  TripDispatcher is its own class. We also created our own error that is a child of the StandardError Class. 
What _composition_ relations exist between classes? | TripDispatcher is the driver code that connects the various classes. 
Describe a decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | We wanted to decide how to handle the situation where no drivers are available, thus we created a NoDriverError that inherited from the StandardError class. 
Give an example of a _template method_ that you implemented for this assignment | self.from_csv(record)
Give an example of a _nominal test_ that you wrote for this assignment | calculating the right float for the average rating given trips that had concluded.
Give an example of an _edge case test_ that you wrote for this assignment | return zero if there are no trips.
What is a concept that you gained more clarity on as you worked on this assignment | object and inheritance.

### CODE REVIEW

File: lib/trip.rb [hunk 1]
```diff
@@ -46,9 +56,20 @@ def inspect
...
-    def connect(passenger)
+    def connect(passenger, driver)
       @passenger = passenger
+      @driver = driver
       passenger.add_trip(self)
+      driver.add_trip(self)
+    end
+
+    def duration
+      if end_time != nil
```
Comment (beccaelenzil) [inline lib/trip.rb:67]: Nice work handling in progress trips across the board (especially testing for this edge case!). I'm not sure if this would make sense with the rest of your implementation, but you might consider whether raising an exception instead of returning a string could work for in progress trips.

File: test/passenger_test.rb [hunk 2]
```diff
@@ -69,6 +75,179 @@
...
+        )
+
+      @passenger.add_trip(trip3)
+    end
+
+    it "Calculates total time spent on trips" do
+      expect(@passenger.total_time_spent).must_equal 4500
+    end
+
+    it "Returns 0 if passenger has no trips" do
```
Comment (beccaelenzil) [inline test/passenger_test.rb:225]: Great work testing for these 2 edge cases

File: lib/driver.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    def average_rating
+      valid_ratings = @trips.reject{|trip|trip.rating == nil}
+      all_ratings = valid_ratings.map{|trip|trip.rating.to_f}
+      return 0 if all_ratings.length == 0
+      average_rating = all_ratings.sum / all_ratings.length
+      return average_rating
+    end
+
+    def total_revenue
+      valid_costs = @trips.reject{|trip|trip.cost == nil}
```
Comment (beccaelenzil) [inline lib/driver.rb:41]: Good use of an enumerable method

File: lib/trip_dispatcher.rb [hunk 4]
```diff
@@ -27,14 +36,50 @@ def inspect
...
+        else 
+          latest_trip = indiv_driver.trips.max_by {|trip| trip.end_time}
+          if latest_trip.end_time < oldest_endtime 
+            oldest_endtime = latest_trip.end_time  
+            driver = indiv_driver
+          end
+        end
+      end
+
+      raise NoDriverError.new("Sorry! There are no available drivers. Please request a new trip.") if driver == nil
```
Comment (beccaelenzil) [inline lib/trip_dispatcher.rb:57]: Good use of a custom exception

File: lib/trip_dispatcher.rb [hunk 5]
```diff
@@ -27,14 +36,50 @@ def inspect
               #{passengers.count} passengers>"
     end
 
+    def request_trip(passenger_id)
+      available_drivers = @drivers.select {|driver|driver.status == :AVAILABLE}
+    
+      driver = nil
+      oldest_endtime = Time.now
+
+      available_drivers.each do |indiv_driver|
```
Comment (beccaelenzil) [inline lib/trip_dispatcher.rb:45]: Good work implementing this logic. You may consider encapsulating this logic in a helper method.

File: test/trip_dispatch_test.rb [hunk 6]
```diff
@@ -121,5 +122,66 @@ def build_test_dispatcher
         end
       end
     end
+
+    describe "request_trip(passenger_id)" do
```
Comment (beccaelenzil) [inline test/trip_dispatch_test.rb:126]: Nice work writing comprehensive tests for this method.

**Comment by beccaelenzil** [comment]
# OO Ride Share



## Major Learning Goals/Code Review



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
Comprehension Questions |Your answer about composition relationship gets at this. By composition we mean a _has-a_ or _has-many_ relationship. Examples in this project include "a TripDispatcher _has-many_ Drivers, Passengers and Trips", "a Driver _has-many_ Trips", "a Trip _has-a_ Driver".
The code demonstrates individual learning about `Time` and the responsibility of `Trip.from_csv`, and uses `Time.parse` in `Trip.from_csv` | ✔️
The code demonstrates breaking out complex logic in helper methods, such as making a helper method in `Trip` to calculate duration | ✔️
There are tests for the nominal cases for the `Passenger#net_expenditures` and `Passenger#total_time_spent` | ✔️
There is at least one edge case test for either `Passenger#net_expenditures` or `Passenger#total_time_spent` testing if the passenger has no trips | ✔️
Practices inheritence. `Driver` inherits from `CsvRecord`, and implements `from_csv` | ✔️
Employs problem-solving and implements `Driver#average_rating` and `Driver#total_revenue` | ✔️
Implements the `TripDispatcher#request_trip`, which creates an instance of `Trip` with a driver and passenger, adds the new trip to `@trips`, and changes the status of the driver | ✔️
Practices composition. In `TripDispatcher#request_trip`, the driver gets connected to the new trip, the passenger gets connected to the new trip | ✔️
Practices git with at least 10 small commits and meaningful commit messages | ✔️

## Testing Requirements

| Testing Requirement | yes/no |
| --- | --- |
There is reasonable test coverage for wave 1, and all wave 1 tests pass | ✔️
There is reasonable test coverage for wave 2, and all wave 2 tests pass | ✔️
Wave 3: Tests in wave 1 and wave 2 explicitly test that only completed trips should be calculated (and ignore in-progress trips) | ✔️
There is reasonable test coverage for `TripDispatcher#request_trip`, and all tests pass | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 8+ in Code Review && 3+ in Functional Requirements | ✔️





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅

**Review by beccaelenzil — COMMENTED** [review]
Great work on this project. The comprehensiveness of your tests is particularly notable. I've left a few in line comments for you to consider, but it is clear that the learning goals around TDD, object composition, and inheritance were met. Keep up the hard work!
